### PR TITLE
feat: add Aws sdk

### PR DESCRIPTION
**Pull request by fmoral2**
#### Proposed Changes ####

- Add Aws sdk abstraction methods for create and delete instances 
- Refactor and improvements related

#### Testing ####

Checklist:
1. If your PR changes anything on or related to Jenkins, run it pointing to your branch to make sure it's okay.

- n/a
2. Verify code lint; we should not have errors.

- ok
3. Update the documentation if needed.

- ok 
4. Update makefile and docker run if adding new tests.

-n/a
5. Run your tests at least 4 times with all configurations needed and possible.
- ok 

6. If needed test with different os types.

-n/a

### CODE REVIEW

File: pkg/aws/aws.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	"github.com/rancher/distros-test-framework/config"
+	"github.com/rancher/distros-test-framework/shared"
+)
+
+type Client struct {
+	c   *config.Infra
```
Comment (rancher-max) [inline pkg/aws/aws.go:18]: rename to `infra` might be a bit more clear

**Review by rancher-max — COMMENTED** [review]
Few comments, overall very nice! Thank you!

File: pkg/aws/aws.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+func (a Client) DeleteInstance(ip string) error {
+	if ip == "" {
+		return shared.ReturnLogError("must sent a ip: %s\n", ip)
+	}
+
+	data := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("ip-address"),
```
[the remark this answers is not in the document]
Comment (fmoral2) [reply pkg/aws/aws.go:117]: yeah, agreed. lets add when needed

File: shared/aux.go [hunk 3]
```diff
@@ -85,8 +86,8 @@ func BasePath() string {
 	return filepath.Join(filepath.Dir(b), "../..")
 }
 
-// EnvDir returns the environment directory of the project based on the package passed.
-func EnvDir(pkg string) (string, error) {
+// envDir returns the environment directory of the project based on the dir package passed.
+func envDir(dir string) (string, error) {
```
[the remark this answers is not in the document]
Comment (fmoral2) [reply shared/aux.go:90]: but then in each pkg you call , you will need to add different path

this is just a nicer user friendly way to do it 
 
and the base path it was already being used in a lot of places reuturning to actual home "../.."

this one is on the root project

File: shared/aux.go [hunk 4]
```diff
@@ -96,18 +97,35 @@ func EnvDir(pkg string) (string, error) {
...
+	switch dir {
 	case "factory":
 		c = filepath.Dir(filepath.Join(callerDir))
 		env = filepath.Join(c, "config/.env")
 	case "entrypoint":
-		c = filepath.Dir(filepath.Join(callerDir, ".."))
+		c = filepath.Dir(filepath.Join(callerDir, "."))
+		env = filepath.Join(c, "config/.env")
+	case "pkg":
+		c = filepath.Dir(filepath.Join(callerDir, "."))
```
[the remark this answers is not in the document]
Comment (fmoral2) [reply shared/aux.go:108]: good call

File: shared/aux.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rancher-max) [reply shared/aux.go:90]: https://github.com/rancher/distros-test-framework/blob/main/shared/aux.go#L81-L82
> // BasePath returns the base path of the project.

I think that is the root of the project also right?

File: pkg/aws/aws.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+	return &Client{
+		c: &config.Infra{
+			Ami:              env.Ami,
+			Region:           env.Region,
+			VolumeSize:       env.VolumeSize,
+			InstanceClass:    env.InstanceClass,
+			Subnets:          env.Subnets,
+			AvailabilityZone: env.AvailabilityZone,
+			SgId:             env.SgId,
+			KeyName:          env.KeyName,
```
[the remark this answers is not in the document]
Comment (rancher-max) [reply pkg/aws/aws.go:47]: All of them!

File: shared/aux.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fmoral2) [reply shared/aux.go:90]: no its not, i will update this doc comment

File: pkg/aws/aws.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (fmoral2) [reply pkg/aws/aws.go:47]: > Can we pull any of the env information from the relevant TFVars as well if it's not available in the ENV? Relevant for `Region` in the aws session above as well

which one would be relevant for Region, that is not here yet?  sorry miss that

File: shared/aux.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rancher-max) [reply shared/aux.go:90]: It should be. https://pkg.go.dev/runtime#Caller
> The argument skip is the number of stack frames to ascend, with 0 identifying the caller of Caller.

Meaning that this returns the string for the file aux.go, then we just need to go up one level to get to the base path. This function today actually has a bug in that it goes up two levels, so it is returning the path directly before `distros-test-framework` but every function that calls it is then appending `/distros-test-framework`.

I think we should actually fix this bug, fix the functions that are using it, and just use this to always get the correct path instead of the extra `case` logic in this `envDir`

File: pkg/aws/aws.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rancher-max) [reply pkg/aws/aws.go:47]: oh sorry, line 33 in this same file!

File: shared/aux.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fmoral2) [reply shared/aux.go]: got it , fixed

File: config/reader.go [hunk 6]
```diff
@@ -9,16 +9,16 @@ import (
 )
 
 var (
-	product *ProductConfig
+	product *Product
 	once    sync.Once
 )
 
-type ProductConfig struct {
+type Product struct {
```
Comment (mdrahman-suse) [inline config/reader.go:16]: Not sure for others but to me these references could cause possible confusions down the line as we have a struct called `Product` and it contains string var called `Product`.. and in SU we discussed about adding a package called `product.go`... wondering if there is any way to incorporate `Product` related stuff together and also come up with better naming conventions

File: factory/config.go [hunk 7]
```diff
@@ -28,6 +29,18 @@ type Cluster struct {
 	NumServers   int
 	NumAgents    int
 	Config       clusterConfig
+	Infra
+}
+
+type Infra struct {
```
Comment (mdrahman-suse) [inline factory/config.go:35]: This struct can be renamed to `EC2Props`.. just a suggestion...

File: config/reader.go (hunk 6 above)
Comment (fmoral2) [reply config/reader.go:16]: lets review this PR only within the scope.
what should we naming the product (rke2,k3s) config then >?

File: factory/config.go (hunk 7 above)
Comment (fmoral2) [reply factory/config.go]: done

**Review by rancher-max — DISMISSED** [review]
LGTM, thank you this is great!

File: factory/config.go [hunk 8]
```diff
@@ -105,14 +105,14 @@ func addClusterConfig(
...
-	c.Infra.InstanceClass = terraform.GetVariableAsStringFromVarFile(g, varDir, "instance_class")
-	c.Infra.Subnets = terraform.GetVariableAsStringFromVarFile(g, varDir, "subnets")
-	c.Infra.AvailabilityZone = terraform.GetVariableAsStringFromVarFile(g, varDir, "availability_zone")
-	c.Infra.SgId = terraform.GetVariableAsStringFromVarFile(g, varDir, "sg_id")
-	c.Infra.KeyName = terraform.GetVariableAsStringFromVarFile(g, varDir, "key_name")
+	c.AwsEc2.Ami = terraform.GetVariableAsStringFromVarFile(g, varDir, "aws_ami")
+	c.AwsEc2.Region = terraform.GetVariableAsStringFromVarFile(g, varDir, "region")
+	c.AwsEc2.VolumeSize = terraform.GetVariableAsStringFromVarFile(g, varDir, "volume_size")
+	c.AwsEc2.InstanceClass = terraform.GetVariableAsStringFromVarFile(g, varDir, "ec2_instance_class")
+	c.AwsEc2.Subnets = terraform.GetVariableAsStringFromVarFile(g, varDir, "subnets")
```
Comment (fmoral2) [inline factory/config.go:112]: ,

File: factory/config.go [hunk 9]
```diff
@@ -105,14 +105,14 @@ func addClusterConfig(
...
-	c.Infra.VolumeSize = terraform.GetVariableAsStringFromVarFile(g, varDir, "volume_size")
-	c.Infra.InstanceClass = terraform.GetVariableAsStringFromVarFile(g, varDir, "instance_class")
-	c.Infra.Subnets = terraform.GetVariableAsStringFromVarFile(g, varDir, "subnets")
-	c.Infra.AvailabilityZone = terraform.GetVariableAsStringFromVarFile(g, varDir, "availability_zone")
-	c.Infra.SgId = terraform.GetVariableAsStringFromVarFile(g, varDir, "sg_id")
-	c.Infra.KeyName = terraform.GetVariableAsStringFromVarFile(g, varDir, "key_name")
+	c.AwsEc2.Ami = terraform.GetVariableAsStringFromVarFile(g, varDir, "aws_ami")
+	c.AwsEc2.Region = terraform.GetVariableAsStringFromVarFile(g, varDir, "region")
+	c.AwsEc2.VolumeSize = terraform.GetVariableAsStringFromVarFile(g, varDir, "volume_size")
+	c.AwsEc2.InstanceClass = terraform.GetVariableAsStringFromVarFile(g, varDir, "ec2_instance_class")
```
Comment (fmoral2) [inline factory/config.go:111]: .

**Review by rancher-max — DISMISSED** [review]

File: config/reader.go (hunk 6 above)
Comment (mdrahman-suse) [reply config/reader.go:16]: Sure we can discuss about naming conventions and possible resolutions in separate PR

**Review by mdrahman-suse — DISMISSED** [review]

**Comment by mdrahman-suse** [comment]
We should move out the file refactoring code to a separate PR

**Review by mdrahman-suse — APPROVED** [review]
LGTM although there are still a lot of refactor done that seems outside the scope of this PR

File: modules/install/join_rke2_master.sh
Comment (aganesh-suse) [inline modules/install/join_rke2_master.sh:1]: The changes to this file do not seem relevant to aws changes. Though nice to be clean with spaces and stuff, let us avoid unncessary rebase needs if we can, and keep a PR clean with only relevant file changes.

File: modules/install/k3s_master.sh
Comment (aganesh-suse) [inline modules/install/k3s_master.sh:1]: https://github.com/rancher/distros-test-framework/pull/44/files#r1414560940

File: modules/install/rke2_master.sh
Comment (aganesh-suse) [inline modules/install/rke2_master.sh:1]: https://github.com/rancher/distros-test-framework/pull/44/files#r1414560940

File: modules/k3s/main.tf
Comment (aganesh-suse) [inline modules/k3s/main.tf:1]: https://github.com/rancher/distros-test-framework/pull/44/files#r1414560940

File: modules/k3s/master/instances_server.tf
Comment (aganesh-suse) [inline modules/k3s/master/instances_server.tf:1]: https://github.com/rancher/distros-test-framework/pull/44/files#r1414560940

File: modules/k3s/worker/instances_worker.tf
Comment (aganesh-suse) [inline modules/k3s/worker/instances_worker.tf:1]: https://github.com/rancher/distros-test-framework/pull/44/files#r1414560940

File: modules/rke2/windows_worker/instances_worker.tf
Comment (aganesh-suse) [inline modules/rke2/windows_worker/instances_worker.tf:1]: https://github.com/rancher/distros-test-framework/pull/44/files#r1414560940

File: modules/k3s/master/policy.yaml [hunk 10]
```diff
@@ -80,7 +80,7 @@ metadata:
 spec:
   podSelector:
     matchLabels:
-      k8s-app: metrics-server
+      k8s-app: metrics-master
```
Comment (aganesh-suse) [inline modules/k3s/master/policy.yaml:83]: was this change intentional - or was this a part of mass replace of 'server' back to 'master' ?

**Comment by aganesh-suse** [comment]
The ProductConfig to Product variable/struct naming related change, EnvConfig() related change, changing the BasePath references/paths/calls... are all not relevant to AWS. Such changes are distracting the actual changes that are needed.

File: modules/install/join_rke2_master.sh
Comment (aganesh-suse) [reply modules/install/join_rke2_master.sh:1]: For example, these space related changes will be more appropriate when we work the rhel fix PR we wanted to do.

File: modules/k3s/master/policy.yaml (hunk 10 above)
Comment (fmoral2) [reply modules/k3s/master/policy.yaml:83]: nope tks

File: modules/install/join_rke2_master.sh
Comment (fmoral2) [reply modules/install/join_rke2_master.sh:1]: done

File: modules/install/join_rke2_master.sh [hunk 11]
```diff
@@ -1,11 +1,11 @@
 #!/bin/bash
-# This script is used to join one or more nodes as masters to the first master
+# This script is used to join one or more nodes as servers to the first server
 echo "$@"
-
-# set -x
-# PS4='+(${LINENO}): '
-# set -e
-# trap 'echo "Error on line $LINENO: $BASH_COMMAND"' ERR
+#
```
[the remark this answers is not in the document]
Comment (fmoral2) [reply modules/install/join_rke2_master.sh:4]: done

File: modules/install/k3s_master.sh
Comment (fmoral2) [reply modules/install/k3s_master.sh:1]: done

File: modules/install/rke2_master.sh
Comment (fmoral2) [reply modules/install/rke2_master.sh:1]: done

**Comment by fmoral2** [comment]
> The ProductConfig to Product variable/struct naming related change, EnvConfig() related change, changing the BasePath references/paths/calls... are all not relevant to AWS. Such changes are distracting from the actual changes that are needed. These could potentially be a refactor PR in itself. Please note that the PR review cycle will be much simpler/shorter, once we start getting only relevant changes into a particular PR.

almost all of this were comments from people that i was addressing.

File: modules/install/k3s_master.sh
Comment (aganesh-suse) [reply modules/install/k3s_master.sh:1]: we are literally changing the exact same lines in this PR: https://github.com/rancher/distros-test-framework/pull/49/files#diff-468d784d1bac451bdb7d6a37dad69394a18f312a129d1d2673e6d706409a41e8R4

And this change is not relevant to aws. Can we please try to avoid multiple conflict fixes for the same lines so many times within the same week.

Let us please remove all unnecessary shell and tf script space/uncomment changes from this PR if possible. Thanks.

File: modules/install/join_rke2_master.sh
Comment (aganesh-suse) [reply modules/install/join_rke2_master.sh:1]: https://github.com/rancher/distros-test-framework/pull/44#discussion_r1415998460

File: modules/install/k3s_master.sh
Comment (fmoral2) [reply modules/install/k3s_master.sh:1]: @aganesh-suse  sorry i got confused with a lot of links and direct links to other codes

can u kindly state here what u want me to change? tks

**Review by VestigeJ — APPROVED** [review]
This is great will be good to push work to AWS directly - you should definitely do a deep dive on using this.

**Review by aganesh-suse — APPROVED** [review]
LGTM. Thanks for fixing everything 👍
